accounting/invoice/utils.py

```diff
@@ -21,27 +21,115 @@
+    >>>     'a': lambda a: a**2
+    >>>     'b': 'str'
+    >>> }
+    >>> resolve_dict_callables(d, 4)
+    >>> print(d)
+    >>> # We lose the callable.
+    >>> {'a': 16, 'b': 'str'}
     """
-    Return the `datetime` object representing the first day of the month that just passed.
-    """
-    return get_last_day_past_month().replace(day=1)
+    new_dict = {}
+    for key, potential_callable in dictionary.items():
+        if callable(potential_callable):
+            new_dict[key] = potential_callable(*args)
+        else:
+            new_dict[key] = dictionary[key]
+    return new_dict
 
 
-def default_invoice_number():
+# pylint: disable=too-many-arguments,too-many-locals
+def send_email_with_invoice(template, subject, extra_email_context=None, client_usernames=settings.BILLING_CYCLE_USERS,
+                            create_invoice=False, draft_invoice=False, fill_line_items_from_jira=False,
+                            upload_to_google_drive=False, auto_approve=False):
     """
-    Return the invoice number which defaults to `yyyy-mm` for the current year and month.
+    Send an email with an invoice attached, and potentially perform peripheral actions.
+
+    TODO: This function's pretty beefy; break it up by encapsulating logic into more utility functions or a new class.
+
+    :param template: The email template.
+    :param subject: The email subject.
+    :param extra_email_context: Extra kwargs to update the email context with.
+    :param client_usernames: The list of billing cycle users' usernames.
+    :param create_invoice: Whether to create a new invoice based off of a past one.
+    :param draft_invoice: Whether this is a draft invoice or not.
+    :param fill_line_items_from_jira: Whether to fill up the invoice's line items from JIRA worklogs.
+    :param upload_to_google_drive: Whether to upload the resulting invoice PDF to Google Drive.
+                                   If using a draft invoice, it gets uploaded into `invoices-in/draft/{month}`.
     """
-    return timezone.now().strftime("%Y-%m")
+    if extra_email_context is None:
+        extra_email_context = {}
+    now = datetime.datetime.now()
+    past = get_last_day_past_month()
+    past_month_formatted = past.strftime('%B')
+    template = get_template(template)
+    for client_username in client_usernames:
+        client = Account.objects.get(user__username=client_username)
+        providers = [rate.provider for rate in client.client_hourly_rates.all()]
+        for provider in providers:
+            # Copy over certain details from the latest invoice from the proper month to the new invoice.
+            invoice = models.Invoice.objects.filter(
+                provider=provider,
+                client=client,
+                date__month=past.month if draft_invoice else now.month,
+            ).latest('date')
+            if create_invoice:
+                invoice = models.Invoice.objects.create(
+                    number=invoice.number,
```

Comment:
@UmanShahzad Shouldn't that number be incremented? The default format is YYYY-MM if I remember correctly? Or YYYY-XX, with XX starting from 01?

Reply:
Hmm, that's true, but that actually presents a new problem I didn't think about. People have totally different schemes for this: some with the default format, some with OC-XX, some with just XX, and so on. There are ways to handle this, like just iterate through a list of possible schemes and see if there's a match, and if so use that scheme's iterating method on the number. I can go through the current schemes that people in the team are using and build this, but it would have to be maintained for new schemes.

Reply:
I bit the bullet and went ahead and did this -- luckily it's not a mess, and in fact very readable code. Will push later.

Reply:
@UmanShahzad Just seeing this comment now - it would be fine to have a simple approach on this for now, ie the most common use-case(s), and do the rest by hand. No need to solve all the problems at once : ) Good that you found a simple solution already though!

Reply:
@antoviaque Yep, I just went ahead and did it, and it'll prevent any further manual work from here on. (One would have to do the manual work no matter what every month otherwise).